src/plist.h

```diff
@@ -26,38 +26,85 @@
 #ifndef GAP_PLIST_H
 #define GAP_PLIST_H
 
+#include <src/debug.h> /* for GAP_ASSERT */
 
 /****************************************************************************
 **
-
 *F  NEW_PLIST(<type>,<plen>)  . . . . . . . . . . . allocate a new plain list
 **
 **  'NEW_PLIST'  allocates    a new plain   list  of  type <type> ('T_PLIST',
 **  'T_SET', 'T_VECTOR') that has room for at least <plen> elements.
 **
-**  Note that 'NEW_PLIST' is a  macro, so do not call  it with arguments that
-**  have side effects.
 */
-#define NEW_PLIST(type,plen)            NewBag(type,((plen)+1)*sizeof(Obj))
+static inline Obj NEW_PLIST(UInt type, Int plen)
+{
+    GAP_ASSERT(plen >= 0);
+    return NewBag(type, (plen + 1) * sizeof(Obj));
+}
 
+/****************************************************************************
+**
+*F  IS_PLIST( <list> )  . . . . . . . . . . . check if <list> is a plain list
+*/
+static inline Int IS_PLIST(Obj list)
+{
+    return FIRST_PLIST_TNUM <= TNUM_OBJ(list) &&
+           TNUM_OBJ(list) <= LAST_PLIST_TNUM;
+}
 
 /****************************************************************************
 **
-*F  GROW_PLIST(<list>,<plen>) . . . .  make sure a plain list is large enough
+*F  IS_PLIST_ASSERT_CHECK( <list> )  . . . . .check if <list> is a plain list
 **
-**  'GROW_PLIST' grows  the plain list <list>  if necessary to ensure that it
-**  has room for at least <plen> elements.
+**  This version of IS_PLIST checks if 'COPYING' is set, and allows
+**  T_POSOBJ and T_DATOBJ (for backwards compatability reasons, as PLIST
+**  methods are often used on these types).
```

Comment:
Aha, I see. In `FuncSTRONGLY_CONNECTED_COMPONENTS_DIGRAPH` it really seems like a bug, and easy enough to fix.
It would be really good to know all the instances of this, and get them fixed (resp. submit patches to the package authors). Of course that can happen after this PR has been merged (and then hopefully eventually we can make this particular check stricter again).